application/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+                call = call_rq.recv() => { call(&mut ()); }
+            };
+        });
+
+        log::info!("Application {} running", self.name);
+
+        // Wait for a subsystem to shut down.
+        self.shutting_down_rx.recv().await.expect("All subsystems already dead");
+
+        // Drop the unused shutdown request end
+        std::mem::drop(self.shutdown_request_rx);
+        // Signal the application is shut down so it does not wait for itself
+        std::mem::drop(self.shutting_down_tx);
+
+        log::info!("Application {} shutting down", self.name);
+
+        // Order all the remaining subsystems to shut down.
+        let _ = self.shutdown_request_tx.send(());
+
+        // Wait for the subsystems to go down.
+        // TODO add shutdown timeout here
+        Self::wait_for_subsystems_to_shut_down(self.shutting_down_rx).await;
+
+        log::info!("Application {} terminated", self.name);
+    }
+}
+
+/// Internal action type sent in the channel.
+type Action<T> = Box<dyn FnOnce(&mut T) + Send>;
+
+/// Call request
+pub struct CallRequest<T>(mpsc::Receiver<Action<T>>);
+
+impl<T: 'static + Send> CallRequest<T> {
+    /// Receive an external call to this subsystem.
+    pub async fn recv(&mut self) -> Action<T> {
+        match self.0.recv().await {
+            // We have a call, return it
+            Some(action) => action,
+            // All handles to this subsystem dropped, suspend call handling.
+            None => std::future::pending().await,
+        }
+    }
+}
+
+/// Shutdown request
+pub struct ShutdownRequest(broadcast::Receiver<()>);
+
+impl ShutdownRequest {
+    /// Receive a shutdown request.
+    pub async fn recv(&mut self) {
+        match self.0.recv().await {
+            Err(broadcast::error::RecvError::Lagged(_)) => {
+                panic!("Multiple shutdown broadcast requests issued")
+            }
+            Err(broadcast::error::RecvError::Closed) => {
+                panic!("Shutdown channel sender closed prematurely")
+            }
+            Ok(()) => (),
+        }
+    }
+}
+
+/// Subsystem handle.
+///
+/// This allows the user to interact with the subsystem from the outside. Currently, it only
+/// supports calling functions on the subsystem.
+#[derive(Clone)]
+pub struct Subsystem<T> {
+    // Send the subsystem stuff to do.
+    action_tx: mpsc::Sender<Action<T>>,
+}
+
+impl<T: Send + 'static> Subsystem<T> {
+    /// Crate a new subsystem handle.
+    fn new(action_tx: mpsc::Sender<Action<T>>) -> Self {
+        Self { action_tx }
+    }
+
+    /// Dispatch a function call to the subsystem
+    pub async fn call_mut<R: Send + 'static>(
+        &self,
+        func: impl FnOnce(&mut T) -> R + Send + 'static,
+    ) -> R {
+        let (rtx, rrx) = oneshot::channel::<R>();
+
+        self.action_tx
+            .send(Box::new(move |subsys| {
+                let result = func(subsys);
+                rtx.send(result).ok().expect("Value return channel closed");
```

Comment:
I'm not sure if panicking is the best way to handle a closed channel, isn't this akin to acquiring a poisoned mutex? I think it would be better if the subsystem that received the call and noticed that the calling subsystem has crashed or whatever, it would initiate a graceful shutdown of itself instead of panicking.

Reply:
I would rather have the caller decide what to do, instead of having the subsystem acting on its behalf. Generally this is the correct thing to do in all multi-threading based on the fork model. Even in C++, you pass exceptions back to the calling thread and then they're thrown when `std::future::get()` is called, because throwing exceptions (or panics) in threads can have unintended consequences.

I think we should tell the caller "we couldn't do this because the channel is closed", which is returned, maybe with an enum, and then the caller can panic if it wants. This will also make debugging much easier.

Reply:
I think in this case, the channel outlives the calculation and sending of the return value, but I am not super confident about that. The issue is what happens if you request a subsystem call while other subsystem requests a shutdown. I will construct some tests doing it and run them through `loom`.

Reply:
That's exactly what happens in P2P: propagate the "channel closed" error to the main task which currently does nothing. I can't think what else we'd do than shutdown the software because unless we implement restarting mechanisms for each subsystem, it's highly unlikely that the program is usable after one of its "microservices" has crashed. So even if P2P doesn't care that chainstate is down because it can still send TXs to mempool, in the grand scheme of things the software is in such a state that it might as well restarted, that's at least how I see. Anyway, whatever we choose to do, I still think panicking here is not the correct approach.

Reply:
> The issue is what happens if you request a subsystem call while other subsystem requests a shutdown

There should never be any race conditions in this design, otherwise we're doing it wrong. That's actually one of the design goals here. The receiver can only process one request at a time. So a shutdown request is received, all subsequent requests will simply be denied. The receiver will then will simply wait for the queue to become empty then shutdown.

Reply:
Calls now return a `Result` so the caller can detect if it's shouting at a zombie subsystem